File: node/02_structured_extraction.mjs
// Extract structured JSON fields from a document using a schema

import { readFileSync } from "fs";

const API_KEY = process.env.DEEPREAD_API_KEY;
const BASE = "https://api.deepread.tech";
const filePath = process.argv[2];

if (!API_KEY) { console.error("Set DEEPREAD_API_KEY"); process.exit(1); }
if (!filePath) { console.error("Usage: node 02_structured_extraction.mjs <invoice.pdf>"); process.exit(1); }

const schema = JSON.stringify({
  type: "object",
  properties: {
    vendor: { type: "string", description: "Company or vendor name" },
    invoice_number: { type: "string", description: "Invoice or receipt number" },
    date: { type: "string", description: "Invoice date" },
    total: { type: "number", description: "Total amount due" },
    line_items: {
      type: "array",
      items: { type: "object", properties: { description: { type: "string" }, amount: { type: "number" } } },
      description: "List of line items",
    },
  },
});

const form = new FormData();
form.append("file", new Blob([readFileSync(filePath)]), filePath.split("/").pop());
form.append("schema", schema);

const { id: jobId } = await fetch(`${BASE}/v1/process`, {
  method: "POST",
  headers: { "X-API-Key": API_KEY },
  body: form,
}).then(r => r.json());

console.log(`Submitted: ${jobId}`);

let delay = 5000;
while (true) {
  await new Promise(r => setTimeout(r, delay));
  const result = await fetch(`${BASE}/v1/jobs/${jobId}`, {
    headers: { "X-API-Key": API_KEY },
  }).then(r => r.json());

  console.log(`  Status: ${result.status}`);

  if (result.status === "completed") {
    console.log("\n--- Extracted Fields ---");
    console.log(JSON.stringify(result.result.data, null, 2));
    break;
  } else if (result.status === "failed") {
    console.error(`Failed: ${result.error}`);
    break;
  }
  delay = Math.min(delay * 1.5, 15000);
}

File: python/02_structured_extraction.py
"""Extract structured JSON fields from a document using a schema."""

import os
import sys
import json
import time
import requests

API_KEY = os.environ.get("DEEPREAD_API_KEY")
BASE = "https://api.deepread.tech"

if not API_KEY:
    print("Set DEEPREAD_API_KEY in your environment or .env file")
    sys.exit(1)

file_path = sys.argv[1] if len(sys.argv) > 1 else None
if not file_path:
    print("Usage: python 02_structured_extraction.py <invoice.pdf>")
    sys.exit(1)

headers = {"X-API-Key": API_KEY}

# Define what fields to extract
schema = json.dumps({
    "type": "object",
    "properties": {
        "vendor": {"type": "string", "description": "Company or vendor name"},
        "invoice_number": {"type": "string", "description": "Invoice or receipt number"},
        "date": {"type": "string", "description": "Invoice date"},
        "total": {"type": "number", "description": "Total amount due"},
        "line_items": {
            "type": "array",
            "items": {
                "type": "object",
                "properties": {
                    "description": {"type": "string"},
                    "amount": {"type": "number"}
                }
            },
            "description": "List of line items"
        }
    }
})

# Submit with schema
with open(file_path, "rb") as f:
    resp = requests.post(
        f"{BASE}/v1/process",
        headers=headers,
        files={"file": f},
        data={"schema": schema},
    )
    resp.raise_for_status()

job = resp.json()
job_id = job["id"]
print(f"Submitted: {job_id}")

# Poll with backoff
delay = 5
while True:
    time.sleep(delay)
    result = requests.get(f"{BASE}/v1/jobs/{job_id}", headers=headers).json()
    status = result["status"]
    print(f"  Status: {status}")

    if status == "completed":
        data = result["result"]["data"]
        print("\n--- Extracted Fields ---")
        print(json.dumps(data, indent=2))

        # Show human-in-the-loop flags
        for field, value in data.items():
            if isinstance(value, dict) and value.get("hil_flag"):
                print(f"\n  Warning: '{field}' flagged for review: {value.get('reason', 'low confidence')}")
        break
    elif status == "failed":
        print(f"Failed: {result.get('error', 'Unknown error')}")
        break

    delay = min(delay * 1.5, 15)
